networkit/cpp/community/ParallelLeiden.cpp

```diff
@@ -0,0 +1,530 @@
+                // This will probably never happen
+                if (resize) {
+                    waitingForResize++;
+                    while (resize) {
+                        std::this_thread::yield();
+                    }
+                    waitingForResize--;
+                }
+                cutWeights.resize(vectorSize);
+                assert(inQueue[u]);
+                index currentCommunity = result[u];
+                double maxDelta = std::numeric_limits<double>::lowest();
+                index bestCommunity = none;
+                double degree = 0;
+                for (auto z : pointers) {
+                    // Reset the clearlist : Set all cutweights to 0 and clear the pointer vector
+                    cutWeights[z] = 0;
+                }
+                pointers.clear();
+
+                graph.forNeighborsOf(u, [&](node neighbor, edgeweight ew) {
+                    index neighborCommunity = result[neighbor];
+                    if (cutWeights[neighborCommunity] == 0) {
+                        pointers.push_back(neighborCommunity);
+                    }
+                    if (u == neighbor) {
+                        degree += ew;
+                    } else {
+                        cutWeights[neighborCommunity] += ew;
+                    }
+                    degree += ew; // keep track of the nodes degree. Loops count twice
+                });
+
+                if (pointers.empty())
+                    continue;
+
+                // Determine Modularity delta for all neighbor communities
+                for (auto community : pointers) {
+                    // "Moving" a node to its current community is pointless
+                    if (community != currentCommunity) {
+                        double delta;
+                        delta = modularityDelta(cutWeights[community], degree,
+                                                communityVolumes[community]);
+                        if (delta > maxDelta) {
+                            maxDelta = delta;
+                            bestCommunity = community;
+                        }
+                    }
+                }
+                double modThreshold = modularityThreshold(
+                    cutWeights[currentCommunity], communityVolumes[currentCommunity], degree);
+
+                if (0 > modThreshold || maxDelta > modThreshold) {
+                    moved[omp_get_thread_num()]++;
+                    if (0 > maxDelta) { // move node to empty community
+                        singleton++;
+                        bestCommunity = upperBound++;
+                        if (bestCommunity >= communityVolumes.capacity()) {
+                            // Chances are this will never happen. Ever. Seriously...
+                            bool expected = false;
+                            if (resize.compare_exchange_strong(expected, true)) {
+                                vectorSize += 10000;
```

Comment:
Maybe store `10000` in a member variable, it's more expressive and you avoid magic numbers.

Reply:
Will do.

Reply:
The entire concurrent resizing logic is a bit hard to read. Can you add a comment why this is safe? I assume that it is because no threads access `cutWeights` and `communityVolumes` while the resize operation is in progress? Why is this guaranteed to hold?

Reply:
cutweights is thread-local and communityvolumes is only resized when all threads have yielded so no access can occur.

Ive added this in the code.

Reply:
Thank you, that clears it up!